ct: cte_track event handler

u3s opened the pull request:
### WHAT
- event handling module for tracking CT execution progress
- displays terminal test summary at the end of execution (suitable for reading GH action test run result without HTML)
- in `live`, `live_details` mode of operation also puts summary of not OK results in terminal during execution
- progress data prepared by `cte_track` can also be redirected to a file with `file` option containing a path
### WHY
- if some test result is not OK, GH action executing OTP tests is typically:
- too big and unreadable
- not displaying any summary at the end
- also sometimes, there is too little info in terminal and having some notification about test progress might be wanted

### USAGE
- provide CT with `cte_track` event handler in any way of your choice
- for GH actions by modifying related test spec, add this: `{event_handler, {cte_track, [{mode, live_details}]}}.`
- for terminal use and testing OTP with make, one may adjust EXTRA_ARGS variable passed to CT used by OTP scripts

### sample output - summary
![image](https://github.com/erlang/otp/assets/4620572/2ee0a72b-bfb7-41d3-8097-73415870c0e1)
### sample output - live information
![image](https://github.com/erlang/otp/assets/4620572/1319b2d2-80d3-4197-a4fa-6536e8bce61b)

jchristgit commented:
This looks good, thanks for putting it up!

I guess one question that we need to figure out: how to deal with the output generated from the test server. For instance, skipped tests:
```erl
    print(major, "=result        ~w: ~tp", [ReportTag,Reason1]),
    print(1, "*** SKIPPED ~ts ***",
	  [get_info_str(Mod,Func, CaseNum, get(test_server_cases))]),
```
The test server (or, at least, `test_server_ctrl`) seems to deal with both HTML and text output.

Perhaps it would be beneficial to separate these out and have the test server deal purely with interpreting individual test results and passing them to whoever is interested (e.g. hooks)?

jchristgit commented:
@u3s with the update to #7375 and the new plan, especially "PR2' - (no dependencies) is hook handler printing expected summary to text file (could be a value on its own and merged quick)", I was wondering whether it would make sense to re-open this pull request? I started working on a hook to provide output when I remembered this pull request is similar to what I would end up with. I guess tests and documentation would be missing, depending on if you have time I can contribute it

u3s commented:
> @u3s with the update to #7375 and the new plan, especially "PR2' - (no dependencies) is hook handler printing expected summary to text file (could be a value on its own and merged quick)", I was wondering whether it would make sense to re-open this pull request? I started working on a hook to provide output when I remembered this pull request is similar to what I would end up with. I guess tests and documentation would be missing, depending on if you have time I can contribute it

great you came back. I mentally got used to thought we will have this feature in OTP-27 ;-).

Yes. I will re-open. I think it has some value even as a standalone feature.

I must be very cautious promising my contribution capacity, but I speculate I should merge it in few weeks.
At least the basic version. We can later adjust it to fit better with `ct_run -out` PR.
Please comment what would do you think should be adjust or added in this basic stage of this PR.

Regarding new plan:
- I think critical thing would be to start `ct_run -out` related work and prioritize checking what worried you most - AFAIR how/if `ct:log` could be redirected to file. Can you take that?

u3s commented:
1. I've experimented with `cte_track` event handling module for that PR. Please have a look and comment.
2. currently the branch is in work-in-progress stage (I wanted to see how it works with cth_log_redirect). I will clean it up later.
3. I hope you can make a PR with your cth_log_redirect update - they could be nicely combined with `cte_track`
4. with `cte_track` and update `cth_log_redirect` - I think we will have some tools for making GH actions test runs more readable (I think this was our original concern.)
5. enabling this could be done per wanted applications in related test specs (I will include few in this PR, probably postpone CT as it turned out more tricky)
6. for `ct_run -out`, I must say honestly this might be tougher than expected and probably not a priority for me in upcoming months :-/. Current thoughts:
    - I also have some 2nd thoughts on the direction and concept
    - I also think test_server knows too much about test reporting - we observed that when fixing `-stylesheet`;
    - maybe (as you suggest) some abstraction could improve the situation - so WHAT is not interleaved with HOW (I mean mixing of test event with generating reports about that)
    - There are border cases which might be tricky (c apps printing to terminal, CT verifying `ct:pal` behavior ... and probably many others
    - I would need much more code reading and prototyping for making educated suggestion on where and how we should proceed - currently I have very limited time for that work and would like to park it for a while

u3s commented:
- branch is clean and ready to merge from my perspective
- @jchristgit do you have any comments or suggestions?

jchristgit reviewed (commented):
This is really nice! :+1:

This is probably my personal preference, but I think `live` is more readable than `live_details`. With `live_details` a group being skipped will print this for every test case, which gets noisy. I would expect `live_details` to also print one-line infos for each passed test case to prevent the skipped notifications from messing up the output.

While running the inets test cases I had some stray "SKIPPED" notifications:
```
.Testing make_test_dir.inets_test: *** SKIPPED test case 403 of 502 ***
... ...........
```
I assume we can't silence these without some change to the test server, but shouldn't this also have printed the "SKIPPED" line from the event handler? Can be reproduced with:

```diff
diff --git a/lib/inets/test/inets.spec b/lib/inets/test/inets.spec
index cf45e0febd..4cc9134cee 100644
--- a/lib/inets/test/inets.spec
+++ b/lib/inets/test/inets.spec
@@ -1,4 +1,4 @@
-{suites,"../inets_test", all}.
+{suites,"../inets_test", [httpd_SUITE]}.
```

Also, at the final summary the counts don't match up, and I'm confused as to what `ct_framework` is doing here:

```sh
ct_framework [.........] (419 ok 0 failed 82 skipped)
inets_SUITE [.......] (426 ok 0 failed 82 skipped)
inets_socketwrap_SUITE [..] (428 ok 0 failed 82 skipped)
inets_sup_SUITE [......] (434 ok 0 failed 82 skipped)
uri_SUITE [..] (436 ok 0 failed 82 skipped)
Testing make_test_dir.inets_test: TEST COMPLETE, 425 ok, 0 failed, 77 skipped of 502 test cases
```

Finally, being pedantic I'm wondering whether maybe another name would fit a bit better? `cte_track`, where does it track? What does the tracking contain? Maybe something like `cte_console_track` or `cte_cli_progress`? Again, this is probably just personal preference.

u3s commented:
> This is probably my personal preference, but I think `live` is more readable than `live_details`. With `live_details` a group being skipped will print this for every test case, which gets noisy. I would expect `live_details` to also print one-line infos for each passed test case to prevent the skipped notifications from messing up the output.

I'm currently thinking about dropping `live_details`.
Instead `live` could be also `{live, [skipped, failed, ok]}` - so one could customize what is printed in live mode. By default only single chars.

I think it will be a longer journey to get cleaner terminals when running tests. For each app situation seems different:
1. ssh - is mainly spammed with logger
2. ssl - skips hundreds of testcases (would be good to get rid of skip notices since it is too many of them)
  - however, in general skips are also of interest because sometimes new skips
3. CT - running run_test and run_script_test generate direct writes to terminal in large amounts; also logger redirection settings must be provided to internally started nodes

> 
> While running the inets test cases I had some stray "SKIPPED" notifications:
> 
> ```
> .Testing make_test_dir.inets_test: *** SKIPPED test case 403 of 502 ***
> ... ...........
> ```
> 
> I assume we can't silence these without some change to the test server, but shouldn't this also have printed the "SKIPPED" line from the event handler? Can be reproduced with:
> 
> ```diff
> diff --git a/lib/inets/test/inets.spec b/lib/inets/test/inets.spec
> index cf45e0febd..4cc9134cee 100644
> --- a/lib/inets/test/inets.spec
> +++ b/lib/inets/test/inets.spec
> @@ -1,4 +1,4 @@
> -{suites,"../inets_test", all}.
> +{suites,"../inets_test", [httpd_SUITE]}.
> ```

Yeah. Probably some bug, I will check.

> Also, at the final summary the counts don't match up, and I'm confused as to what `ct_framework` is doing here:
> 
> ```shell
> ct_framework [.........] (419 ok 0 failed 82 skipped)
> inets_SUITE [.......] (426 ok 0 failed 82 skipped)
> inets_socketwrap_SUITE [..] (428 ok 0 failed 82 skipped)
> inets_sup_SUITE [......] (434 ok 0 failed 82 skipped)
> uri_SUITE [..] (436 ok 0 failed 82 skipped)
> Testing make_test_dir.inets_test: TEST COMPLETE, 425 ok, 0 failed, 77 skipped of 502 test cases
> ```

Yes. This is kind of expected to me ;-) That is why i changed title to "EVENTS SUMMARY" ;-).
In CT, probably from historical reasons (test_server -> CT evolution and other work), we have a small terminology issue around functions in test suite module.
In some contexts (more external), we split them into: test functions and config functions (init_per_suite, etc.)
In other contexts (more internal), we treat all functions from test suite module the same (referring to them as `tc`, which might be confused with test function, since looking like acronym of a testcase).

To an amateur code archeologist (me ;-)), it looks like new(smaller) definition of testcase was introduced at some point.

Note: events `tc_start`, `tc_done` look the same for test and config functions.

Treating all functions equally is good for troubleshooting IMHO because it exposes internal flow CT executed and is more informative. Maybe numbers printed could somehow printed with division for test and config functions ... I don't know how to do that cleanly ... do you have some idea?

As long as we don't document it I hope it will be fine, to have this dichotomy - it is not this PR that introduced it after all.

> Finally, being pedantic I'm wondering whether maybe another name would fit a bit better? `cte_track`, where does it track? What does the tracking contain? Maybe something like `cte_console_track` or `cte_cli_progress`? Again, this is probably just personal preference.

I will think about it more. I had no better idea, then I picked something I would like to memorize and type.

jchristgit commented:
> I'm currently thinking about dropping `live_details`.
> Instead `live` could be also `{live, [skipped, failed, ok]}` - so one could customize what is printed in live mode. By default only single chars.

This sounds good.

> I think it will be a longer journey to get cleaner terminals when running tests. For each app situation seems different:
>   [...]

Yep, especially in CT I agree there is definitely some work to be done
to get it to be quieter. But Rome was not built in a day :-)

> Yes. This is kind of expected to me ;-) That is why i changed title to "EVENTS SUMMARY" ;-).
>
>   [...]
> 
> Treating all functions equally is good for troubleshooting IMHO because it exposes internal flow CT executed and is more informative. Maybe numbers printed could somehow printed with division for test and config functions ... I don't know how to do that cleanly ... do you have some idea?

Hm, an interesting piece of code history...

I am 50/50 on this. I agree having the internal flow is more informative
and could help with debugging, but personally when I think of a "test
case" I think of just a test function and not the scaffolding around it.
On the other hand if we were to exclude config functions, it raises the
question how the hook would display failures in them?

I would say: at least `ct_framework` should be hidden in the output. And
as for displaying it, maybe we could count the "config" functions in a
separate "config" number? But personally I would say "hide them unless
they crash".

I think someone new to CT would be confused how the counts match up or
what ct_framework is doing there, but if it's not documented for now,
it's not a problem.

u3s commented:
sorry for delay. I hope to come back to this soon.

u3s commented on lib/common_test/src/cte_track.erl:
```diff
@@ -0,0 +1,227 @@
+    {ok, State};
+handle_event(#event{name = tc_done, data = Data},
+             State = #{device := D}) ->
+    print(D, "~n~nUnandled interesting event:~nName = tc_done~nData = ~p~n~n",
```
fixed

u3s commented:
- i've rebased branch
- simplified code to have only basic stuff for now
- would like to stay with `cte_track` name - to me module is tracking CT events
- merge it soon, so I can test it more while doing other stuff
- for now I would treat it as internal code and not document it until it is good enough

u3s commented:
PR is merged to master. I treat it as 1st step.
At least now the mess in "Run tests" github actions is scrollable and more informative.

Further cleanup is assumed - now more interesting changes might be required.
If we came up with better name, we can do it later.
